# Applying Advanced Transformations 
- Priscilla Cole
- 11/19/2022

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os, json

# Part 1 (From LP)
Clean the files and combine them into one final DataFrame.
This dataframe should have the following columns:

- Hero (Just the name of the Hero)
- Publisher
- Gender
- Eye color
- Race
- Hair color
- Height (numeric)
- Skin color
- Alignment
- Weight (numeric)
- Plus, one-hot-encoded columns for every power that appears in the dataset. E.g.: (will be done at the end)

- Agility
- Flight
- Superspeed
- etc. 

In [2]:
# Load in super hero info DF
df_i = pd.read_csv('Data/superhero_info - superhero_info.csv')
df_i.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [3]:
# Load in super hero powers DF
df_p = pd.read_csv('Data/superhero_powers - superhero_powers.csv')
df_p.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [4]:
# check for duplicates
df_i.duplicated().sum()

0

In [5]:
df_p.duplicated().sum()

0

In [6]:
# check for missing data
df_i.isna().sum()

Hero|Publisher    0
Gender            0
Race              0
Alignment         0
Hair color        0
Eye color         0
Skin color        0
Measurements      0
dtype: int64

In [7]:
df_p.isna().sum()

hero_names    0
Powers        0
dtype: int64

In [8]:
# check general info
df_i.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Hero|Publisher  463 non-null    object
 1   Gender          463 non-null    object
 2   Race            463 non-null    object
 3   Alignment       463 non-null    object
 4   Hair color      463 non-null    object
 5   Eye color       463 non-null    object
 6   Skin color      463 non-null    object
 7   Measurements    463 non-null    object
dtypes: object(8)
memory usage: 29.1+ KB


In [9]:
df_p.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   hero_names  667 non-null    object
 1   Powers      667 non-null    object
dtypes: object(2)
memory usage: 10.5+ KB


In [10]:
# split columns in first df
df_i['Hero|Publisher'].head()

0            A-Bomb|Marvel Comics
1    Abe Sapien|Dark Horse Comics
2              Abin Sur|DC Comics
3       Abomination|Marvel Comics
4     Absorbing Man|Marvel Comics
Name: Hero|Publisher, dtype: object

In [11]:
# Need to split on the "|" using expand = True
df_i['Hero|Publisher'].str.split('|', expand = True)

0                  1
0             A-Bomb      Marvel Comics
1         Abe Sapien  Dark Horse Comics
2           Abin Sur          DC Comics
3        Abomination      Marvel Comics
4      Absorbing Man      Marvel Comics
..               ...                ...
458     Yellowjacket      Marvel Comics
459  Yellowjacket II      Marvel Comics
460             Yoda       George Lucas
461          Zatanna          DC Comics
462             Zoom          DC Comics

[463 rows x 2 columns]

In [12]:
# Save the new columns into the dataframe
df_i[['Hero', 'Publisher']] = df_i['Hero|Publisher'].str.split('|',
                                                                expand = True)
df_i.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

In [13]:
# Drop first Hero|Publisher column
df_i = df_i.drop(columns = 'Hero|Publisher')
df_i.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

In [14]:
# Examine Measurments column
measure = df_i.loc[0,"Measurements"]
print(type(measure))
measure

<class 'str'>


"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

In [15]:
# Need to approach this like a str, and not a dict (test)
measure = measure.replace("'", '"') # Look closely, this is hard to read
measure

'{"Height": "203.0 cm", "Weight": "441.0 kg"}'

In [16]:
# json.loads (test)
fixed_measure = json.loads(measure)
print(type(fixed_measure))
fixed_measure

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [17]:
# apply to column

df_i['Measurements'] = df_i['Measurements'].str.replace("'", '"')

# Now for Json.loads to the full column
df_i['Measurements'] = df_i['Measurements'].apply(json.loads)

df_i['Measurements'].head()

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
Name: Measurements, dtype: object

In [18]:
# Check a value
test_measure = df_i.loc[0, 'Measurements']
print(type(test_measure))
test_measure

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [19]:
measurements = df_i['Measurements'].apply(pd.Series)
measurements

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

In [20]:
df_i = pd.concat((df_i, measurements), axis = 1)
df_i.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher    Height    Weight  
0      Marvel Comics  203.0 cm  441.0 kg  
1  Dark Horse Comics  191.0 cm   65.0 kg  
2          DC Comics  185.0 cm   90.0 kg  
3      Marvel Comics  203.0 cm  441.0 kg  
4      Marvel Comics  193.0 cm  122.0 kg

In [21]:
# Drop first 'Measurements' column
df_i = df_i.drop(columns = 'Measurements')

In [22]:
# check to make sure
df_i.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher    Height    Weight  
0         A-Bomb      Marvel Comics  203.0 cm  441.0 kg  
1     Abe Sapien  Dark Horse Comics  191.0 cm   65.0 kg  
2       Abin Sur          DC Comics  185.0 cm   90.0 kg  
3    Abomination      Marvel Comics  203.0 cm  441.0 kg  
4  Absorbing Man      Marvel Comics  193.0 cm  122.0 kg

In [23]:
# Expand "Height" and "Weight" for just the numeric values
df_i['Height'].str.split(' ', expand = True)

0   1
0    203.0  cm
1    191.0  cm
2    185.0  cm
3    203.0  cm
4    193.0  cm
..     ...  ..
458  183.0  cm
459  165.0  cm
460   66.0  cm
461  170.0  cm
462  185.0  cm

[463 rows x 2 columns]

In [24]:
# Save new columns into the dataframe
df_i[['Height(cm)', 'CM']] = df_i['Height'].str.split(' ', expand = True)
df_i.head(3)

Gender           Race Alignment Hair color Eye color Skin color        Hero  \
0   Male          Human      good    No Hair    yellow    Unknown      A-Bomb   
1   Male  Icthyo Sapien      good    No Hair      blue       blue  Abe Sapien   
2   Male        Ungaran      good    No Hair      blue        red    Abin Sur   

           Publisher    Height    Weight Height(cm)  CM  
0      Marvel Comics  203.0 cm  441.0 kg      203.0  cm  
1  Dark Horse Comics  191.0 cm   65.0 kg      191.0  cm  
2          DC Comics  185.0 cm   90.0 kg      185.0  cm

In [25]:
# Weight
df_i[['Weight(kg)', 'KG']] = df_i['Weight'].str.split(' ', expand = True)
df_i.head(3)

Gender           Race Alignment Hair color Eye color Skin color        Hero  \
0   Male          Human      good    No Hair    yellow    Unknown      A-Bomb   
1   Male  Icthyo Sapien      good    No Hair      blue       blue  Abe Sapien   
2   Male        Ungaran      good    No Hair      blue        red    Abin Sur   

           Publisher    Height    Weight Height(cm)  CM Weight(kg)  KG  
0      Marvel Comics  203.0 cm  441.0 kg      203.0  cm      441.0  kg  
1  Dark Horse Comics  191.0 cm   65.0 kg      191.0  cm       65.0  kg  
2          DC Comics  185.0 cm   90.0 kg      185.0  cm       90.0  kg

In [26]:
# Drop 'CM', 'KG', 'Height', and 'Weight' columns from dataframe
df_i = df_i.drop(columns = ['CM', 'KG', 'Height', 'Weight'])
df_i.head(2)

Gender           Race Alignment Hair color Eye color Skin color        Hero  \
0   Male          Human      good    No Hair    yellow    Unknown      A-Bomb   
1   Male  Icthyo Sapien      good    No Hair      blue       blue  Abe Sapien   

           Publisher Height(cm) Weight(kg)  
0      Marvel Comics      203.0      441.0  
1  Dark Horse Comics      191.0       65.0

In [27]:
# Change Heigh(cm) and Weight(kg) to floats data type
df_i['Height(cm)'] = df_i['Height(cm)'].astype(float)
df_i['Weight(kg)'] = df_i['Weight(kg)'].astype(float)
df_i.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Gender      463 non-null    object 
 1   Race        463 non-null    object 
 2   Alignment   463 non-null    object 
 3   Hair color  463 non-null    object 
 4   Eye color   463 non-null    object 
 5   Skin color  463 non-null    object 
 6   Hero        463 non-null    object 
 7   Publisher   463 non-null    object 
 8   Height(cm)  463 non-null    float64
 9   Weight(kg)  463 non-null    float64
dtypes: float64(2), object(8)
memory usage: 36.3+ KB


In [28]:
df_p.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [29]:
df_p['powers_split'] = df_p['Powers'].str.split(",")
df_p.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        powers_split  
0    [Agility, Super Strength, Stamina, Super Speed]  
1  [Accelerated Healing, Durability, Longevity, S...  
2  [Agility, Accelerated Healing, Cold Resistance...  
3                               [Lantern Power Ring]  
4  [Accelerated Healing, Intelligence, Super Stre...

In [30]:
# Exploding column of lists
exploded = df_p.explode('powers_split')
exploded[['hero_names', 'Powers', 'powers_split']].head(5)

hero_names                                             Powers  \
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   

          powers_split  
0              Agility  
0       Super Strength  
0              Stamina  
0          Super Speed  
1  Accelerated Healing

In [31]:
# Saving values from the exploded column
cols_to_make = exploded['powers_split'].dropna().unique()
cols_to_make

array(['Agility', 'Super Strength', 'Stamina', 'Super Speed',
       'Accelerated Healing', 'Durability', 'Longevity', 'Camouflage',
       'Self-Sustenance', 'Cold Resistance', 'Underwater breathing',
       'Marksmanship', 'Weapons Master', 'Intelligence', 'Telepathy',
       'Immortality', 'Reflexes', 'Enhanced Sight', 'Sub-Mariner',
       'Lantern Power Ring', 'Invulnerability', 'Animation',
       'Super Breath', 'Dimensional Awareness', 'Flight', 'Size Changing',
       'Teleportation', 'Magic', 'Dimensional Travel',
       'Molecular Manipulation', 'Energy Manipulation', 'Power Cosmic',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Heat Resistance',
       'Matter Absorption', 'Regeneration', 'Stealth', 'Power Suit',
       'Energy Blasts', 'Energy Beams', 'Heat Generation', 'Danger Sense',
       'Phasing', 'Force Fields', 'Hypnokinesis', 'Invisibility',
       'Enhanced Senses', 'Jump', 'Shapeshifting', 'Elasticity',
 

In [32]:
for col in cols_to_make:
    df_p[col] = df_p['Powers'].str.contains(col)

df_p.head(3)

/var/folders/4m/z2mf66p530vgp2ngq2nnw59c0000gn/T/ipykernel_11322/982269933.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df_p[col] = df_p['Powers'].str.contains(col)
/var/folders/4m/z2mf66p530vgp2ngq2nnw59c0000gn/T/ipykernel_11322/982269933.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df_p[col] = df_p['Powers'].str.contains(col)
/var/folders/4m/z2mf66p530vgp2ngq2nnw59c0000gn/T/ipykernel_11322/982269933.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` man

/var/folders/4m/z2mf66p530vgp2ngq2nnw59c0000gn/T/ipykernel_11322/982269933.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df_p[col] = df_p['Powers'].str.contains(col)
/var/folders/4m/z2mf66p530vgp2ngq2nnw59c0000gn/T/ipykernel_11322/982269933.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df_p[col] = df_p['Powers'].str.contains(col)
/var/folders/4m/z2mf66p530vgp2ngq2nnw59c0000gn/T/ipykernel_11322/982269933.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` man

hero_names                                             Powers  \
0     3-D Man         Agility,Super Strength,Stamina,Super Speed   
1      A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2  Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   

                                        powers_split  Agility  Super Strength  \
0    [Agility, Super Strength, Stamina, Super Speed]     True            True   
1  [Accelerated Healing, Durability, Longevity, S...    False            True   
2  [Agility, Accelerated Healing, Cold Resistance...     True            True   

   Stamina  Super Speed  Accelerated Healing  Durability  Longevity  ...  \
0     True         True                False       False      False  ...   
1     True        False                 True        True       True  ...   
2     True        False                 True        True       True  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  

[3 rows x 170 columns]

In [33]:
# Merging and changing column names
hero_df = pd.merge(df_i, df_p, left_on = 'Hero', right_on = 'hero_names')

In [34]:
hero_df.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher  Height(cm)  Weight(kg)  ...  \
0         A-Bomb      Marvel Comics       203.0       441.0  ...   
1     Abe Sapien  Dark Horse Comics       191.0        65.0  ...   
2       Abin Sur          DC Comics       185.0        90.0  ...   
3    Abomination      Marvel Comics       203.0       441.0  ...   
4  Absorbing Man      Marvel Comics       193.0       122.0  ...   

  Weather Control Omnipresent Omniscient  Hair Manipulation  Nova Force  \
0           False       False      False              False       False   
1           False       False      False              False       False   
2           False       False      False              False       False   
3           False       False      False              False       False   
4           False       False      False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 180 columns]

In [35]:
hero_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 463 entries, 0 to 462
Columns: 180 entries, Gender to Changing Armor
dtypes: bool(167), float64(2), object(11)
memory usage: 126.1+ KB


In [36]:
# drop columns
hero_df = hero_df.drop(columns = ['hero_names', 'Powers', 'powers_split'])

In [37]:
# Average weight for all heroes
hero_df['Weight(kg)'].mean()

113.58963282937366

In [38]:
# Locate our heroes with super speed
super_speed = hero_df.loc[hero_df['Super Speed'] == True]

In [39]:
# Average weight of the heroes with super speed
super_speed['Weight(kg)'].mean()

129.40404040404042

In [40]:
# Remove super speed
no_super_speed = hero_df.loc[hero_df['Super Speed'] == False]

In [41]:
no_super_speed['Weight(kg)'].mean()

101.77358490566037

- Average Weight with Super Speed = 129.4
- Average Weight without Super Speed = 101.8

In [42]:
print(f"Average Weight of Hero's with Super Speed: {super_speed['Weight(kg)'].mean()}")
print(f"Average Weight of Hero's without Super Speed: {no_super_speed['Weight(kg)'].mean()}")

Average Weight of Hero's with Super Speed: 129.40404040404042
Average Weight of Hero's without Super Speed: 101.77358490566037


In [43]:
hero_df['Publisher'].value_counts()

Marvel Comics        297
DC Comics            138
Dark Horse Comics     11
George Lucas           5
Team Epic TV           4
Shueisha               4
Star Trek              2
Unknown                1
Image Comics           1
Name: Publisher, dtype: int64

In [44]:
hero_df.groupby('Publisher')['Weight(kg)'].mean()

Publisher
DC Comics            104.188406
Dark Horse Comics    101.818182
George Lucas          77.400000
Image Comics         405.000000
Marvel Comics        119.579125
Shueisha              64.500000
Star Trek             79.000000
Team Epic TV          72.000000
Unknown               83.000000
Name: Weight(kg), dtype: float64